### PR TITLE
Feature/kakfa exporter - kafka key by TraceID

### PR DESCRIPTION
**Pull request by arik-dig**
**Description:**
Kafka Exporter - been modified so it publish kafka messages with key of TraceID - it will allow to partition the kafka Topic and consume it concurrently

**Link to tracking Issue:** 
relates to issue #12318

**Testing:**
I've installed the modified version and seeing the Key being filled as expected (seeing via Kafka UI)

image

here are entries BEFORE the change (the key does not exist)

image

**Documentation:**
Did not touch it

### CODE REVIEW

**Review by MovieStoreGuy — COMMENTED** [review]
Thank you making the changes, the only concern I have is now we are splitting the traces by default which is a change in behaviour. 

Ideally this should be controlled by a configuration field, something like:
```yaml
key_data: [none, traceid] # OneOf none Trace ID
```

Otherwise, you will need to:

- [ ] Add a change log entry
- [ ] Add some tests around the changes

**Comment by arik-dig** [comment]
@MovieStoreGuy I've updated the PR with `key_data` with default of `none`
can you please re-review?

**Comment by arik-dig** [comment]
@MovieStoreGuy can you please review?

**Review by MovieStoreGuy — COMMENTED** [review]
The idea is good, and I've left comments for things that need to be added. 

I am not entirely convinced on the approach, the way I envision this is there is a new type named `KeyedEncoder` that extends the existing encoding types.

I wonder if it would be possible to do something like:

```go 
type KeyedTracesMarshaler struct {
  base   TraceMarshaler
  keyBy  KeyByFunc
}

func (keyed KeyedTracesMarshaler) Marshal(in ptrace.Traces, topic string) (...) {
  for _, trace := range keyed.keyBy(in) {
     msg, err := keyed.base.Marshal(trace, topic)
     if err != nil {
        return nil, err
     }
     msgs = append(msgs, 
        &sarama.ProducerMessage{
          Topic: topic, 
           Value: sarama.byteEncoder(msg), 
           Key: sarama.ByteEncoder(keyed.getKey(trace)),
       },
     )  
  }
}
```

This would allow you just wrap any existing TraceMarshler and still implement the expected interface without having to extend it.

File: exporter/kafkaexporter/factory.go [hunk 1]
```diff
@@ -118,6 +120,14 @@ type kafkaExporterFactory struct {
 	logsMarshalers    map[string]LogsMarshaler
 }
 
+func KeyOfTracerMarshaller(marshaler TracesMarshaler) string {
```
[the remark this answers is not in the document]
Comment (arik-dig) [reply exporter/kafkaexporter/factory.go:123]: the key used to be only the `Encoding`, see line 48, now the key is combination of `Encoding` and `KeyData`

**Comment by MarcinGinszt** [comment]
Hey! My company needs this feature. @arik-dig, can I take over and apply the maintainers comments?

**Comment by jwafle** [comment]
Hi @arik-dig, @MarcinGinszt!

My company is also looking for this functionality 😆, and fairly urgently . 

As such, I have been working on a new branch [here](https://github.com/jwafle/opentelemetry-collector-contrib/tree/feat/partition-by-trace-id) with an updated implementation and some more robust unit tests. It still needs a bit of cleanup, but I'm planning on opening a PR soon.

Regards.

**Comment by jwafle** [comment]
Linking out the open PR to resolve the same issue: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/29660

**Pull request closed without merging**